### PR TITLE
Superellipsoid

### LINKED ISSUE #294: Superellipsoid
Super-ellipsoid is a very broad class of shapes that should be easy to implement. Something like
```
-shape superellipsoid <r> <t>
```

For references:
* https://en.wikipedia.org/wiki/Superellipsoid
* https://github.com/adda-team/adda/wiki/AddingShape

This shape has been recently used in Kanngießer F. and Kahnert M. Modeling optical properties of non-cubical sea-salt particles, [_J. Geophys. Res. Atmos._ **126**, e2020JD033674](http://doi.org/https://doi.org/10.1029/2020JD033674) (2021).

### PR DESCRIPTION
**Pull request by jeromefung**
### Implementation of superellipsoid shape
Issue #294 indicates interest in adding superellipsoids to the shapes supported by ADDA. I have implemented this shape following the instructions in [the wiki](https://github.com/adda-team/adda/wiki/AddingShape). Because computing the volume of a superellipsoid requires special functions (the gamma and beta functions) not previously implemented in ADDA, I implemented the necessary functions in superellipsoid.c (similar to the implementation of Chebyshev particles). In addition, I modified the Makefile to also compile superellipsoid.c and added tests of the new shape.

The conventions and notation I used are those given in T. Wriedt, "Using the T-Matrix Method for Light Scattering Computations by Non-axisymmetric Particles: Superellipsoids and Realistically Shaped Particles," _Part. Part. Syst. Charact._ **19** (2002), 256-258.

The code explicitly handles the edge cases when the exponent parameters e and n are zero. However, there could be numerical instability for very small nonzero values. Also, large values of e and n result in sharply spiked shapes that may be poorly handled by the DDA.

### Related issues
Fixes #294.

### Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist

- [x] I have read the [contributing guidelines](https://github.com/adda-team/adda/blob/master/.github/CONTRIBUTING.md)
- [x] Code compiles correctly
- [x] Tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works. And these tests pass
- [ ] I have added/extended necessary documentation (if appropriate)

### Further comments
I locally tested the code in several ways. First, a superellipsoid can describe several shapes already present in ADDA: sphere, ellipsoid, box, and cylinder. As can be seen in run_tests.sh below, I ran ADDA for the pre-existing shapes and for a superellipsoid with the corresponding parameters. I then used diff to compare the geometry files as well as the cross sections. The output (test_comparisons) shows no differences other than the names of the shapes.

Second, to check the shapes produced by the new code more generally, I generated geometry files for several values of e and n (including both convex and concave shapes) and visualized the geometry files with LiteBil. The results are summarized in litebil_renderings.pdf.
[litebil_renderings.pdf](https://github.com/adda-team/adda/files/6871946/litebil_renderings.pdf)
[run_tests.sh.txt](https://github.com/adda-team/adda/files/6871950/run_tests.sh.txt)
[test_comparisons.txt](https://github.com/adda-team/adda/files/6871951/test_comparisons.txt)

### CODE REVIEW

**Comment by alkichigin** [comment]
Maxim Yurkin is currently on vacation, he will review this PR in the middle of August

File: src/superellipsoid.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    t = x+g-0.5;
+
+    return sqrt(TWO_PI) * pow(t,x-0.5) * exp(-t) * result;
+  }
+}
+
+//======================================================================================================================
+
+static double Beta(double x,double y)
```
Comment (myurkin) [inline src/superellipsoid.c:85]: Is it possible to make it robust for arbitrary arguments (say, 1000), including avoiding overflow? For instance, using logarithm of gamma function, which should be almost as easy to compute as gamma itself.
Note, that we need one-time computation of Beta, so robustness is preferable to speed (and also to have one function for all possible future needs).

File: src/superellipsoid.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  double nterm,eterm;
+
+  // lim_{n -> 0} n B(n/2 + 1, n) = 1, and lim_{e -> 0} e B(e/2, e/2) = 4.
+  // But direct calculation doesn't work if n and e are too small.
+  // Below a threshold, set them to limiting values.
+  // The threshold is lower than what is probably practical for DDA calculations
+  // since we use finite-size dipoles.
+
+  const double veps = 1e-5;
+  if (n<veps) nterm = 1;
```
Comment (myurkin) [inline src/superellipsoid.c:108]: Related to the above. If calculation of Beta is fully robust, then it should work for small arguments as well. Even if some threshold is used here, we should be sure that double accuracy is guaranteed (series expansion can be used to estimate the proper threshold).

File: src/make_particle.c [hunk 3]
```diff
@@ -1810,6 +1814,42 @@ void InitShape(void)
...
+	 case SH_SUPERELLIPSOID: {
+		 double aspectY, aspectZ;
+
+		 // Read and test parameters
+		 aspectY=sh_pars[0];
+		 TestPositive(aspectY,"aspect ratio y/x");
+		 aspectZ=sh_pars[1];
+		 TestPositive(aspectZ,"aspect ratio z/x");
+		 se_e=sh_pars[2];
+		 TestRangeII(se_e,"superellipsoid exponent e",0,10);
```
Comment (myurkin) [inline src/make_particle.c:1826]: It seems that there are no reasons to disallow e and n larger than 10. They do have problem with discretization of the scatterer (may result in zero dipoles), but that is natural. We only need to ensure that calculation of shape parameters work fine.

File: src/make_particle.c [hunk 4]
```diff
@@ -2202,6 +2244,41 @@ void MakeParticle(void)
...
+				  // Cross sections along z are rectangles (superellipse with e=0)
+					// but with a and b scaled according to z.
+					oneOverN=1/se_n;
+					// the scale factor
+					tmp1=pow(1 - pow(4*invsqZ*zr*zr,oneOverN),0.5*se_n);
+					if (fabs(xr)<=0.5*tmp1 && fabs(yr)<=haspY*tmp1 && fabs(zr)<=haspZ) mat=0;
+				}
+				else {
+					// The usual case with n, e > 0.
+					// Singular behavior for very small n and e is still possible!
```
Comment (myurkin) [inline src/make_particle.c:2271]: Can this be improved?

**Review by myurkin — COMMENTED** [review]
Dear Jerome. Thanks a lot for taking the initiative and carefully preparing the pull request. And sorry for taking a long time to review. Everything seems to be working fine. There are only several small things to comply with the code style (see inline comments). Some of other changes I have just committed to this branch (please also review them).

Apart from that there are a few questions related to robustness of computations (for all rare cases). If you can, please do it as well. Otherwise, I will take a look at the next iterations.

I also like the script for tests that you created - I will think for a place, where it should reside (maybe a bit different framework to existing `tests/2exec`).

**Comment by myurkin** [comment]
Additional note concerning the gamma and beta functions. I recommend to use `lgamma` or `tgamma` functions available in C99 standard. Using `lgamma` seems most robust (especially for positive arguments), for instance B(x,y) would be exp(lgamma(x)+lgamma(y)-lgamma(x+y)). The latter will probably even work for very small x and y.

Then we probably do not need a separate source (`superellipsoid.c`), instead a small piece of code in `make_particle.c` would suffice.

**Comment by myurkin** [comment]
I have also committed the updated manual (`.docx`). Please check the additions corresponding to superellipsoid shape.

If you can, please also make a representative illustration of superellipsoid to be added to Fig.3. Importantly, it should be in the same style as existing ones. The svg sources of the latter (e.g., for InkScape) can be found in `/doc/figs/shapes.svg`.

**Comment by jeromefung** [comment]
Hi Maxim -- your addition to the manual looks good to me. 

I'm going to take a look at `lgamma` and refactor (or get rid of) `superellipsoid.c` and test the changes. I may not be able to get to this until later in the week.

I can also draw an svg figure for the superellipsoid shape.

File: src/make_particle.c [hunk 5]
```diff
@@ -1608,6 +1611,7 @@ void InitShape(void)
 			}
 			hdratio=diskratio/2;
 			volume_ratio=PI_OVER_FOUR*diskratio;
+			printf("%f\n",volume_ratio);
```
[the remark this answers is not in the document]
Comment (jeromefung) [reply src/make_particle.c]: Yes, this got left in by mistake. Fixed by most recent commit.

File: src/make_particle.c [hunk 6]
```diff
@@ -1810,6 +1814,42 @@ void InitShape(void)
 			Nmat_need=2;
 			break;
 		}
+	 case SH_SUPERELLIPSOID: {
```
[the remark this answers is not in the document]
Comment (jeromefung) [reply src/make_particle.c]: Fixed in latest commit.

[the remark this answers is not in the document]
Comment (myurkin) [reply src/make_particle.c]: The alignment still seems to be off (compared to other cases above and below), but let's first finalize the logic of the code.

**Comment by myurkin** [comment]
Thanks, Jerome. Take your time and let me know when you finish everything that you can. I will take it from there.

**Comment by myurkin** [comment]
Jerome, sorry for the delay. There were some issues with small arguments. For instance, the following command lines failed:
Small e with non-zero n always leads to a box
`adda -shape superellipsoid 1 1 0.001 1`

Small n with non-zero e always leads to a void shape
`adda -shape superellipsoid 1 1 1 0.001`

Anyway, I have fixed it and now the generation seems to be completely robust. I have also added tests to `/tests/equiv` based on the file you uploaded here. Now it tests all kind of crazy arguments to superellipsoid shape.

Concerning the manual, I have re-implemented your changes starting with my original file. The .docx is not very reliable when edited on many different systems (some invisible changes in document structure, etc. can crawl in). And I also updated the figure of shapes inside the manual (thanks for your svg art). If you want to add any other changes to the manual, please describe them in a comment.

So please test the updated code, including using the `/tests/equiv/superellipsoid.sh`. Once you approve it, I will merge.

**Comment by jeromefung** [comment]
Sorry for the delayed response. I've looked at the changes to the code, including the comparison tests in `/tests/equiv/superellipsoid.sh`, and everything looks good to me. I think this can be merged.

Generating the superellipsoid shape for the svg was tricky because I was trying to draw a 3D-projected wireframe shape defined by mathematical functions. While not the most elegant solution, I did what I found easiest: I wrote a Python program to calculate the points to be drawn and write them to a file, rendered the drawing using LaTeX and TikZ, and then used the [pdf2svg](https://github.com/dawbarton/pdf2svg) utility to convert the rendering to svg. If you envision needing to modify the actual superellipsoid drawing in the figure in the future, I'm happy to add the relevant Python and tex source code to the repository in some appropriate place.

**Comment by myurkin** [comment]
Jerome, thanks again for your efforts in submitting and finalizing this pull request. And congratulations with the first merged pull request for ADDA. Welcome to the club - https://github.com/adda-team/adda/graphs/contributors

**Pull request merged**